### DISCUSSION TITLE
Proposal #4: DIDComm Protocol as the basis for a Trust Spanning Layer Protocol

### DISCUSSION THREAD

**Discussion by mwherman2000** [trustoverip/trust-spanning-protocol#27] [Ideas]
**[Proposal 4 to ToIP TSL TF 0.2: Web 7.0 Trust Spanning Layer Framework++](https://hyperonomy.com/2023/02/22/proposal-4-to-toip-tsl-tf-web-7-0trust-spanning-layerframework/)**
(https://hyperonomy.com/2023/02/22/proposal-4-to-toip-tsl-tf-web-7-0trust-spanning-layerframework/)

is a summary presentation of the conversations that have been taking place here:

- [Proposal 4: Web 7.0 TSLF Proposal as a Picture: Separate (but parallel) Inbound and Outbound Message Processing Stacks](https://github.com/trustoverip/trust-spanning-protocol/discussions/21) 
- [Proposal 4 Addendum: Separating the intent of Layer 2 from the intentions of Layer 3](https://github.com/trustoverip/trust-spanning-protocol/discussions/23) 

Cheers, 
Michael Herman (Web 7.0)

**Comment by mwherman2000** [comment]


**Reply by neiljthomson** [reply]
Link to the _ToIP Layer 2 Trusted Spanning Protocol Assessment Checklist_ document?

**Comment by mwherman2000** [comment]
I'm continually updating and refining the Proposal 4 presentation. I've now updated the presentation to version 0.10.

Primary changes:

- Clarified the meanings and application of: (with examples)
    - Layer 3 Super Protocols 
    - Layer 2 Base (Trust Spanning) Protocol
    - Layer 1 Subprotocols
- Several other small changes

Version 0.10 can be found at the same URL: https://hyperonomy.com/2023/02/22/proposal-4-to-toip-tsl-tf-web-7-0trust-spanning-layerframework/

**Comment by mwherman2000** [comment]
Version 0.14 of Proposal 4 presents a compelling story (with irrefutable evidence and examples) supporting the selection of the DID Communications (DIDComm) protocol as the _basis_ for the Layer 2 Trust Spanning Protocol for any/all decentralized system ecosystems:
- Web 7.0 DIDComm Architecture Reference Model
- ToIP Technical Architecture Specification (ToIP TAS), 
- etc.

Please enjoy the following video podcast and downloadable presentation. I look forward to your comments and feedback.

Version 0.14 of Proposal 4 (video podcast and downloadable presentation) can be found at the same/usual URL:

https://hyperonomy.com/2023/02/22/proposal-4-to-toip-tsl-tf-web-7-0trust-spanning-layerframework/

NOTE: The above video podcast references version 0.11 of Proposal 4.

NOTE: Version 0.14 of the presentation added a small number of feature changes/additions; most notably, the addition of several Layer 1 Subprotocol and Layer 3 Super Protocol use cases including the Issuer-Holder-Verifier model (starting near slide 51).

![image](https://user-images.githubusercontent.com/6101736/222720237-daf1074e-cbd6-4550-b8ba-ef5860dd7a4e.png)

**Reply by talltree** [reply]
@mwherman2000 I reviewed the slides and watched your Version 0.11 video podcast. Once I understood your nomenclature and iconography, I think I understood everything, and I agree with your many compelling use cases and examples.

As you summarize at the end, your proposal is that we should "select the DID Communications (DIDComm) protocol as the Layer 2 Trust Spanning Protocol for any/all decentralized system ecosystems". However in [Proposal #2](https://github.com/trustoverip/trust-spanning-protocol/discussions/19), @dhh1128, one of the architects of DIDComm V2, did not propose that we should use it as currently specified. My understanding is that he believes some changes should be made to optimize it for the design requirements of the trust spanning layer protocol. Have you discussed that with him?

**Reply by dhh1128** [reply]
I believe that DIDComm v2 ticks all the boxes, and I like Michael's proposal because I think it makes that clear. I can hardly say otherwise, since it's something I poured my heart and soul into. However, that does not mean I think it is optimal, which is why I didn't just recommend it in its current form.

DIDComm v2 relies on ECDH-1PU for authcrypt. This algorithm is an IETF Draft RFC, but it has never got past the draft stage, and it is currently stale, not headed for standardization. That mean that adopting/implementing is a little bit less straightforward than I would like (although SICPA successfully contributed support for it in nimbus [a mainstream, globally used java JOSE library] and in one or two other major JOSE libraries for other programming languages -- so this isn't a total dealbreaker). Critics of DIDComm who say it is using exotic crypto aren't accurate, but it is not wrong to say that DIDComm could benefit from slightly more mainstream approaches.

KERI's approach is to sign everything. At first I was resistant to this because I felt that non-repudiation was problematic for privacy. However, Sam convinced me that the unit of repudiation is an identifier, and we can always create new identifiers to isolate reputation in one conversation from reputation in another. Thus, signing is NOT problematic for privacy. And it eliminates the need for authcrypt, making the crypto more efficient, more mainstream, and actually harder to crack as well. So a version of DIDComm with KERI-style signing underneath it is better, IMO, than what DIDComm v2 has.

KERI also introduces robust support for multisig. DIDComm can benefit from this. DIDComm can be used with (almost) any DID method, but using ones that support multisig is dramatically better for security. It allows the correct modeling of identity control in organizations, among other things. So DIDComm on top of KERI is, IMO, significantly better than DIDComm by itself, in this respect.

KERI also introduces a prerotation scheme that is a huge security innovation. IMO this ought to be used very broadly. I very much want our TSP to have that feature.

KERI also formalizes self-certifying identifiers in a very useful way. Peer DIDs, which DIDComm has always been enthusiastic about, have this important security characterstic. However, KERI is a more elegant and robust implementation of the idea, and provides an upgrade path for people who do want to use blockchains (of any stripe) in addition to those who don't.

(Despite the virtues of KERI, I don't believe it's minimally sufficient, by itself. I think we cannot transfer trust from one trust task to another without some of DIDComm's constructs, for example...)

For all of these reasons, I think that DIDComm by itself, while very good and useful, should not be where the conversation ends. The core of DIDComm, integrated with KERI, feels like something even better to me. That's why my proposal was a little more nuanced than proposal 4. But they have a lot in common. :-)

**Reply by Willem-de-kok** [reply]
@dhh1128 two questions for clarification. 

1. In this sentence "However, Sam convinced me that the unit of repudiation is an identifier, and we can always create new identifiers to isolate reputation in one conversation from reputation in another. " Do you mean repudiation where you wrote reputation?

2. Could you clarify what is means that 'the unit of repudiation is an identifier'? And why this means that signing is not a problem for privacy?
At first I interpreted it as: A certain key is associated with the identifier (in other words is attributed to an identifier). Consider the case where the sender of a message controls (as can be shown using KERI protocols) the identifier. My interpretation: non-repudiation means that it cannot be denied by the sender that the sender's identifier has been provided with proof of delivery of the sent message, i.e. the delivery proof was processed by the sender's identifier. However, since the sender has a key event log, the sender can still show that he did/didn't receive the message, based on the key event log, which would solve the problem of privacy.
But, after some rethinking, I think this interpretation is incorrect. I think you mean that the problem of privacy would occur if you used the same identifier in communication with everyone. Then, you could easily be bother with all kinds of messages once that identifier is known to a scammer and you would always have affirmation of receival processed (non-repudation). This could be circumvented by creating different identifiers (probably together with different keys you sign the identifier with) for different conversations. Is this a correct interpretation?

**Reply by talltree** [reply]
@Willem-de-kok Having been part of one long in-person conversation between @dhh1128 and @SmithSamuelM on this very question, I'm 90% sure your latter interpretation is correct. 

In other words, a sender can control one aspect of the privacy of a message by controlling which identifier the sender uses to send the message. For example, a message the sender wishes to attribute to the sender's public or legal identity can use a publicly verifiable identifier for the sender. But a message that the senders wishes to send pseudonymously can use a cryptographically verifiable identifier generated by the sender for use as a one-time pseudonym. Or anything in between, e.g., a sender can also use a groupwise pseudonym.

In all cases, the messages are signed so that they are cryptographically verifiable and non-repudiable by the sender. But by controlling which type of identifier is used, the sender can control the degree of privacy that message has with regard to its ability to be associated with the sender's real-world identity.

**Reply by dhh1128** [reply]
>In this sentence "However, Sam convinced me that the unit of repudiation is an identifier, and we can always create new identifiers to isolate reputation in one conversation from reputation in another. " Do you mean repudiation where you wrote reputation?

No, I meant what I typed. Repudiation is an attempt to dissociate an action/statement from one's reputation. If Alice wants to associate statement X with her reputation as a gamer, she uses the keys associated with Alice.gamer-identifier. If she wants to associate statement X with her reputation as an employee, she uses the keys associated with Alice.employee-identifier. As long as Alice does not provide correlators between these two, they remain entirely separate. Alice.employee-identifier can repudiate anything done by Alice.gamer-identifier, and vice versa. If Alice has a new context where she wants to have no connection to her personal reputation in any previous contexts, she simply creates a new pseudonym.

**Reply by Willem-de-kok** [reply]
Thank you for your explanations! The relation between repudiation and reputation is a lot clearer. And I can see how this contributes to privacy in a positive way.

**Reply by mwherman2000** [reply]
> As you summarize at the end, your proposal is that we should "select the DID Communications (DIDComm) protocol as the Layer 2 Trust Spanning Protocol for any/all decentralized system ecosystems".

@talltree I'm going to stick with my recommendation that the DID Communications (DIDComm) protocol server as the **basis** for the Layer 2 Trust Spanning Layer. Based on @dhh1128 comments above, I believe he also supports using  the DIDComm protocol as a _basis_ - albeit with some additional refinements/improvements.  

As of _today_, the DIDComm protocols meets all of the requirements stipulated by ToIP for a Layer 2 Trust Spanning Protocol and goes well beyond.  My Proposal 4 also "goes well beyond" in making a compelling (irrefuatable?) presentation for the DIDComm protocol to be the basis for the Layer 2 Trust Spanning Protocol.  Some of the previous proposal presentations very much lack this latter quality.

![image](https://user-images.githubusercontent.com/6101736/224468504-ccde8cd6-7b6f-4867-bce6-eb09581fc1cf.png)

![image](https://user-images.githubusercontent.com/6101736/224468562-fd62c5cb-ffde-4fc8-9fe4-66214434ca8f.png)

**Reply by mwherman2000** [reply]
> For all of these reasons, I think that DIDComm by itself, while very good and useful, should not be where the conversation ends. The core of DIDComm, integrated with KERI, feels like something even better to me. That's why my proposal was a little more nuanced than proposal 4. But they have a lot in common. :-)

A key (lethal?) issue for KERI is that over the past 2 or so years, I have yet to consume what I consider to be a _truly understandable and digestible_ presentation of "how" KERI works at a detailed level. I have watched a lot of meetings and recordings as well as in-person meetings at IIW34. I get the very basics but I'm a smart guy and for the most part, 90% of what I hear and see describing KERI is unintelligiable. This past week's re-presentation of Proposal 1 is another example of this IMO.

p.s. If I could grasp it, I would love to be able to produce my style of clear, developer-focused content for KERI - which would benefit a lot of folks.

**Reply by Willem-de-kok** [reply]
@mwherman2000 [this presentation](https://docs.google.com/presentation/d/1lpzYcPrIox9V4hERtn4Kcf7uq01OVU9u3PuVm1aYzR0/edit#slide=id.ga411be7e84_0_18) might be useful.

**Reply by mwherman2000** [reply]
There's 51 slides in this deck.  Which ones are you specifically refering to?  
I've probably watched/listened/read a dozen previous Keri presentations.

**Comment by mwherman2000** [comment]
Version 0.17 of Proposal 4 presents a compelling story (with irrefutable evidence and examples) supporting the selection of the DID Communications (DIDComm) protocol as the _basis_ for the Layer 2 Trust Spanning Protocol for any/all decentralized system ecosystems:
- Web 7.0 DIDComm Architecture Reference Model
- ToIP Technical Architecture Specification (ToIP TAS), 
- etc.

Please enjoy the following video podcast and downloadable presentation. I look forward to your comments and feedback.

Version 0.17 of Proposal 4 (video podcast and downloadable presentation) can be found at the same/usual URL:

https://hyperonomy.com/2023/02/22/proposal-4-to-toip-tsl-tf-web-7-0trust-spanning-layerframework/

NOTE: The above video podcast references version 0.11 of Proposal 4 (the most current version of the video postcast produced to date).

NOTE: Version 0.17 of the presentation added a small number of feature changes/additions; most notably, the addition of Layer 3 Message Processing Overlays and Layer 3 Protocol Overlays (Protocol Specializations) on slides 46, 47, and 48 (73).

![Web 7 0 TSLF Proposal 4-ToIP 0 17-noaudio](https://user-images.githubusercontent.com/6101736/224322085-a0824075-aebb-405f-b22a-eefa7e5eb3d0.png)

**Comment by mwherman2000** [comment]
The final version (version 0.26) of today's Proposal 4 contribution and presentation can be found here: https://hyperonomy.com/2023/02/22/proposal-4-to-toip-tsl-tf-web-7-0trust-spanning-layerframework/

- Proposal 4 webcast to the ToIP TSLP TF NA/EU meeting: https://zoom.us/rec/share/vGaCQMp4xITsAVKsxqv3H1sURX5ZXWd-LjFd2yMaj-PNZlG8rsAyzfdWOZDqEBQ7.55xZUhhvqKIkSZCV
- Proposal 4 webcast  to the ToIP TSLP TF APAC meeting: https://zoom.us/rec/share/IgAXxwX57hj_gGYmKt7qykFh4wuobFJEZuVka5G-dGrxt0axUzZlXpyebqjMuR7c.qLHz7yAF8FILimZw
- Proposal 4 presentation (and contribution to ToIP): https://hyperonomy.files.wordpress.com/2023/03/web-7.0-tslf-proposal-4-toip-0.26-finalx2.pdf (old version. see below for updated versions)
- Proposal 4 GitHub discussion thread: https://github.com/trustoverip/trust-spanning-protocol/discussions/27 (this thread)

**Comment by mwherman2000** [comment]
Version 0.27 updates/change log: Version 0.27 incorporates some changes based on feedback from yesterday's 2 presentations to the ToIP TSL Task Force.  The "FINAL 2" version 0.27 of the presentation can be found in the usual spot: https://hyperonomy.com/2023/02/22/proposal-4-to-toip-tsl-tf-web-7-0trust-spanning-layerframework/

Cc: @wenjing

![image](https://user-images.githubusercontent.com/6101736/225611592-981e2d54-33e7-4a22-bf42-f54d2abca243.png)

![image](https://user-images.githubusercontent.com/6101736/225611621-5bb82ccd-5714-4945-94d2-72882d1b633e.png)

![image](https://user-images.githubusercontent.com/6101736/225685948-9ec1abcf-453b-4f2f-ad65-25ef0e48198a.png)

**Comment by mwherman2000** [comment]
NOTE: The "Implementation View" inspired by Joe Spencer strongly suggests that only 3 layers are required in the Trust Spanning Layer Framework:
- Trust Agent Layer [4]
- Trust Spanning Layer 2
- Trust Foundation Layer 1

- Layer 1 and Layer 2 are connected by a family of TSL Subprotocols. 
- Later 2 and Layer [4] are connected by families of TSL Super Protocols.
- TSL Subprotocols and TSL Super Protocols are _protocol overlays_ built on top of the TSL Base Protocol (DIDComm Messaging V2 and the Credential Sender-Reciever Pattern)

![225611621-5bb82ccd-5714-4945-94d2-72882d1b633e](https://user-images.githubusercontent.com/6101736/225901733-5670bb62-579f-42a0-8d21-8f1e1f880394.png)

![image](https://user-images.githubusercontent.com/6101736/225685948-9ec1abcf-453b-4f2f-ad65-25ef0e48198a.png)

Cc: @dhh1128 @wenjing @talltree

**Comment by mwherman2000** [comment]
Version 0.28 updates/change log: Version 0.28 adds Appendix B. See below. The "FINAL 3" version 0.28 of the presentation can be found in the usual spot: https://hyperonomy.com/2023/02/22/proposal-4-to-toip-tsl-tf-web-7-0trust-spanning-layerframework/

NOTE: If you want to create your own version of the Appendix B slide 106, the above page also contains a link to a downloadable version of the DIDComm Notation graphical design language for decentralized systems.

Appendix B is a partial response to this Proposal 3 discussion thread: https://github.com/trustoverip/trust-spanning-protocol/discussions/25#discussioncomment-5341665

![image](https://user-images.githubusercontent.com/6101736/226189470-35f72625-c984-4aa9-8797-9fae9a05675a.png)

![image](https://user-images.githubusercontent.com/6101736/226189559-fefd7015-a403-428e-ba9a-559021a879f5.png)

Cc: @dhh1128 @wenjing @talltree
